client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +806,128 @@ where
 	}
 }
 
+/// Helper struct for deciding if core part of maintenance procedure shall be executed.
+/// It is publicly exposed only for tests purposes (due to cyclic deps sc_transaction_pool and
+/// substrate_test_runtime_transaction_pool).
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactment was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactment was already performed in f(C1), we should not retract
+/// finalized block)
+/// f(C1), f(C2), nbb(C1) -> false
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+/// nbb(B1), nbb(B2) -> true
+/// nbb(B1), nbb(C1), f(C1) -> false (handle_enactment was already performed in nbb(B1)
+/// nbb(C1), f(B1) -> false (handle_enactment was already performed in nbb(B2)
+#[doc(hidden)]
+pub struct EnactmentState<Block>
+where
+	Block: BlockT,
+{
+	recent_best_block: Option<Block::Hash>,
+	recent_finalized_block: Option<Block::Hash>,
```

Comment:
Do we really need to track both of them? Shouldn't it be enough to just track the last processed block?

We can do a tree route from the last processed to the new block we are about to process. Then we find out about retracted/enacted.

When we have the case that we first received the finalized and later the new best block event, we could skip the new best block event if `block_number_of_new_best <= last_processed_block`?

Reply:
I believe we need to track both of them because we have to distinguish finalized and new best block.

Retracting of finalized blocks would be possible in case of out of order `NBB` events.
But we should not retract finalized block - this is assumption I made while improving `txpool` API.
I did not figure out algorithm with single block kept in state.

Sample:
```
  B1-C1-D1-E1
 /
A
 \
  B2-C2-D2-E2
  ```
- `F(D1)`, `NBB(D2)` -- we don't want to retract D1

Reply:
Good point.